File: backend-express/src/background/handlers/emails.ts
import { Job } from "bullmq";

import { NewsletterData, TASK_NEWSLETTER } from "../setup";

// Create Tasks

async function sendNewsletterTask(job: Job<NewsletterData>): Promise<void> {
    await new Promise((resolve) => setTimeout(resolve, 2000));
    const { name, email } = job.data;
    console.log(
        `Newsletter Email sent to ${name} at following address: ${email}`
    );
}

// Tasks Router
type emailTaskData = NewsletterData;

export async function emailTasksRouter(job: Job<emailTaskData>): Promise<void> {
    switch (job.name) {
        case TASK_NEWSLETTER:
            await sendNewsletterTask(job);
            break;
        default:
            throw new Error(`Unknown job name: ${job.name}`);
    }
}

File: backend-gin/internal/background/handlers/emails.go
package handlers

import (
	"backend/internal/background/tasks/taskspec"
	"context"
	"encoding/json"
	"fmt"

	"github.com/hibiken/asynq"
)

func HandleSendingNewsletter(ctx context.Context, t *asynq.Task) error {
	var data taskspec.NewsletterData
	if err := json.Unmarshal(t.Payload(), &data); err != nil {
		return fmt.Errorf("could not unmarshal data for newsletter task: %w", err)
	}

	message := fmt.Sprintf(
		"Newsletter Email sent to %s at following address: %s",
		data.Name,
		data.Email,
	)
	fmt.Println(message)
	return nil
}
